### PR TITLE
Add manala diff command

### PR DESCRIPTION
**Pull request by ogizanagi**
> https://github.com/manala/manala/pull/37 was improbably closed (missclick) in the exact same time I pushed force the branch. This prevents us from re-opening the previous PR...
### Description

The `manala diff` command generate the diff with the targeted project and manala env templates.

screenshot 2016-10-05 a 00 31 52
### Usage:

``` sh
# inside the manalized project with git (repo must be init):
manala diff
manala diff -q | git patch
```

screenshot 2016-10-04 a 22 57 28
#### Full:

``` sh
manala diff tests/fixtures/Command/DiffTest/AcmeProject/ --env=symfony
```

If saved in a file, patch can be applied by using PHPStorm ( ❤️ ) or by executing:

``` sh
 git patch my_patch.patch
```

inside the working directory.
### Known issues

File mode changes cannot be applied by the `patch` util. It means that using the next `manala diff` after applying the changes will still show file mode changes, and calling `patch -p1` again will return:

> patch: ***\* Only garbage was found in the patch input.

That should not be an issue by using `git apply`.

### CODE REVIEW

File: src/Command/Diff.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $cwd = realpath($input->getArgument('cwd'));
+        $envType = EnvEnum::create($input->getOption('env'));
+
+        if (!is_dir($cwd)) {
+            throw new \RuntimeException(sprintf('The working directory "%s" doesn\'t exist.', $cwd));
+        }
+
+        $originalTemplatePath = MANALA_DIR.'/src/Resources/'.$envType;
```
Comment (chalasr) [inline src/Command/Diff.php]: > A new EnvResourcesLocator might be useful for this. Or simply a new method on the EnvType.

Indeed 👍  some existing code and #29 will need it as well so let's add it in another PR

File: .php_cs [hunk 2]
```diff
@@ -12,6 +12,7 @@ EOF;
 
 $finder = Symfony\CS\Finder\DefaultFinder::create()
     ->in(array(__DIR__.'/src', __DIR__.'/tests'))
+    ->exclude(array(__DIR__.'/tests/fixtures/'))
```
Comment (chalasr) [inline .php_cs]: no need for this change anymore :)

Comment (ogizanagi) [reply .php_cs]: May be required for other fixtures later...but as it doesn't work right now ¯\_(ツ)_/¯

**Comment by chalasr** [comment]
Looks really nice!

**Comment by ogizanagi** [comment]
This one is ready to me. PR description updated. Looking for reviews :)

However, I wonder if we shouldn't set a fake env (not usable with `manala setup`) in order to ease the test fixtures maintenance. WDYT ?

**Comment by chalasr** [comment]
Imho, it is the right way.  I thought about `venv` when @xavier-rdo worked on #19, it might be useful for fully testing it. But maybe you have something else in mind?

**Comment by ogizanagi** [comment]
@chalasr : I'm simply talking about something like this: https://github.com/manala/manala/pull/38/commits/7282583ca2a20a4504f191b69debcfbafe22c645

The `symfony` env will continue to evolve, but the `acme` one only has few reasons to do so and can act as a "sandbox" for our tests.
But we can probably found a mechanism to exclude it from normal usages, at least once the phar is build, until we have a dynamic way to manage envs.

**Comment by chalasr** [comment]
Ok, sorry about the... :trollface: => 🚪 
For sure, it will widely ease maintaining it! 👍

File: .travis.yml [hunk 3]
```diff
@@ -17,8 +18,10 @@ matrix:
 cache:
     directories:
         - $HOME/.composer/cache
-
```
Comment (chalasr) [inline .travis.yml:20]: Could you please revert this change? :)

File: .travis.yml [hunk 4]
```diff
@@ -17,8 +18,10 @@ matrix:
...
 before_install:
+    - sudo add-apt-repository -y ppa:git-core/ppa
+    - sudo apt-get update && sudo apt-get install git-man && sudo apt-get install -y git
+    - git --version
```
Comment (chalasr) [inline .travis.yml:25]: Maybe this was for debugging purpose? Otherwise keeping it isn't a problem to me.

File: src/Exception/InvalidEnvException.php [hunk 5]
```diff
@@ -18,9 +18,9 @@ class InvalidEnvException extends \InvalidArgumentException
     public function __construct($invalidEnvName)
     {
         parent::__construct(sprintf(
-            'The env "%s" doesn\'t exist. Possible values: [%s]',
+            'The env "%s" doesn\'t exist. Possible values: %s',
             $invalidEnvName,
-            implode(' ', EnvEnum::getPossibleEnvs())
+            json_encode(EnvEnum::getPossibleEnvs())
```
Comment (ogizanagi) [inline src/Exception/InvalidEnvException.php:23]: ⚠️ Wild suggestion in the middle 😄

Comment (chalasr) [reply src/Exception/InvalidEnvException.php:23]: 👍

File: src/Resources/acme/ansible/group_vars/app.yml [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  ##############
+  # Supervisor #
+  ##############
+
+  supervisor_configs:
+    - file:     app.conf
+      template: configs/app_program.{{ env }}.j2
+      config:
+        - workflow-worker:
+          - command:        php bin/console app:workflow:worker --run-max=1000 --shutdown-on-exception --no-interaction -vv
```
Comment (chalasr) [inline src/Resources/acme/ansible/group_vars/app.yml]: Damn... This comes from dry*a, and not only this part is custom, meaning that the real template contains the same 😓, I have to fix it after merging this one

File: src/Env/EnvEnum.php [hunk 7]
```diff
@@ -20,6 +20,7 @@
  */
 class EnvEnum
 {
+    const ACME = 'acme'; // This env is not meant to be used except for testing purposes.
```
Comment (chalasr) [inline src/Env/EnvEnum.php]: We should try to find a way to avoid making this env available outside of testing at time to fetch env templates remotely

File: tests/Functional/DiffTest.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+namespace Manala\Tests\Functional;
+
+use Manala\Command\Diff;
+use Manala\Env\EnvEnum;
+use Symfony\Component\Console\Tester\CommandTester;
+
+class DiffTest extends \PHPUnit_Framework_TestCase
+{
+    const ACME_PROJECT_WD = MANALA_DIR.'/tests/fixtures/Command/DiffTest/AcmeProject';
+    const EXPECTED_OUTPUT_FILE = MANALA_DIR.'/tests/fixtures/Command/DiffTest/expected_acme.patch';
```
Comment (chalasr) [inline tests/Functional/DiffTest.php]: A less expressive but simpler way to test this could be to create a temporary `symfony/framework-standard-edition` project (see [BuildTest](https://github.com/manala/manala/blob/master/tests/Functional/BuildTest.php#L38)), manalize it and edit some ansible files, giving an always predictible diff

File: src/Handler/Diff.php [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+    private function copyToTmpLocation($templatePath)
+    {
+        $tmpPath = sys_get_temp_dir().'/Manala';
+
+        if ($this->fs->exists($tmpPath)) {
+            $this->fs->remove($tmpPath);
+        }
+
+        $this->fs->mkdir($tmpPath);
```
Comment (chalasr) [inline src/Handler/Diff.php:137]: Instead of using your own iterator, wouldn't `$this->fs->mirror($templatePath, $tmpPath);` work?

File: src/Command/Diff.php [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        if (!$handler->hasDiff()) {
+            $errIo->success('No diff found.');
+
+            return 0;
+        }
+
+        $errIo->note(<<<'EOTXT'
+You can apply the patch by executing the following command in the project working directory:
+
+$ patch -p1 < my_patch.patch
```
Comment (chalasr) [inline src/Command/Diff.php]: How is set the patch filename? Is it `my_patch.patch`? I miss something ^^
Also imho we should be more explicit about that the project configuration is outdated, maybe something like:

```
Your project configuration is outdated, you can upgrade it by applying the patch on your project:

$ patch -p1 < my_patch.patch
```

**Review by chalasr — APPROVED** [review]
Left minor comments 👍

File: .travis.yml (hunk 4 above)
Comment (ogizanagi) [reply .travis.yml:25]: I left it in order to be sure there is no version with a git version later.

File: .travis.yml (hunk 3 above)
Comment (ogizanagi) [reply .travis.yml:20]: Sure

File: tests/Functional/DiffTest.php (hunk 8 above)
Comment (ogizanagi) [reply tests/Functional/DiffTest.php]: As soon as we always install the exact same version.

I thought about that, but I judge it was too much for the need and will slow down the test suite, and it doesn't solve the `acme` env issue (still required).

File: src/Handler/Diff.php (hunk 9 above)
Comment (ogizanagi) [reply src/Handler/Diff.php:137]: That's probably the `Filesystem` method I haven't found. 😅 
Thanks, I'll give a try.

Comment (ogizanagi) [reply src/Handler/Diff.php:137]: Working great 👍 

Don't know how I missed this method :)

File: tests/Functional/DiffTest.php (hunk 8 above)
Comment (chalasr) [reply tests/Functional/DiffTest.php]: > As soon as we always install the exact same version.

Not sure to understand here, we manalize the temporary project with the same codebase as the one used to check the diff, so the pack is always up to date on both sides, right?

Also I find worst to have an unexisting env in our code base, but let's keep it (we have to quickly add the ability to load remote packs in order to remove it ^^)

File: src/Command/Diff.php (hunk 10 above)
Comment (ogizanagi) [reply src/Command/Diff.php]: > How is set the patch filename? Is it my_patch.patch? I miss something ^^

See https://github.com/manala/manala/pull/38/files#diff-45cbdbae2d21a8c8fae7d15a87e1b135R46

I finally choose to remove the mechanism I used to output the note on stderr, and simply added the text in the command help.

Thus, the command will only output the diff, and that's all. This also allow to use the command simply in quiet mode in order to get the exit code and know if there is a diff or not (0 or 1).

> we should be more explicit about that the project configuration is outdated

IMHO, having a diff is explicit enough.
I've only kept a `success` message to stderr in case there is no diff.

Comment (chalasr) [reply src/Command/Diff.php]: Agreed, the diff directly given to the output is just perfect

**Comment by ogizanagi** [comment]
@chalasr : Should we consider it ready and merge until we have the setup metadata feature ?
IMO it is already usable as it is.

**Comment by chalasr** [comment]
@ogizanagi Great job 👍 Merged. Let's improve it later
